src/bt.rs

```diff
@@ -74,324 +73,217 @@ pub enum BinaryTreeError<V> {
-    /// - `Err(InsertNonEmptyNode(value))` when the subtree already contains a
+    /// - `Err(InsertNonEmptyNode(value))` when the tree already contains a
     /// value at the end of the path.
     /// - `Err(UnreachablePath(value))` when the end of the path is unreachable.
     ///
     /// In the error cases, no insertion occurs and the value is returned to
     /// the caller of this function.
     pub fn insert(&mut self, path: &Path, value: V) -> Result<(), BinaryTreeError<V>> {
-        enum Ref<'a, V> {
-            This(&'a mut Node<V>),
-            Other(&'a mut Option<Box<Node<V>>>),
-        }
-
-        // Finds the node at the end of the path.
-        let mut node = Ref::This(self);
+        let last = self.nodes.len();
+        let mut node = &mut self.root;
         for bit in path.iter() {
-            node = match node {
-                Ref::This(n) => Ref::Other(if !bit { &mut n.left } else { &mut n.right }),
-                Ref::Other(Some(n)) => Ref::Other(if !bit { &mut n.left } else { &mut n.right }),
-                Ref::Other(None) => {
-                    return Err(BinaryTreeError::UnreachableNode(value));
+            match *node {
+                None => return Err(BinaryTreeError::UnreachableNode(value)),
+                Some(next) => {
+                    let n = &mut self.nodes[next];
+                    node = if !bit { &mut n.left } else { &mut n.right };
                 }
-            };
+            }
         }
 
-        // Checks whether the node already has a value,
-        // If so, returns an error wrapping the value that could not be inserted.
-        // otherwise, a new node is created containing the value to be inserted.
-        match node {
-            Ref::This(_) | Ref::Other(Some(_)) => {
-                return Err(BinaryTreeError::InsertNonEmptyNode(value))
-            }
-            Ref::Other(empty_node) => {
-                *empty_node = Some(Box::new(Node::new(value)));
+        if node.is_some() {
+            return Err(BinaryTreeError::InsertNonEmptyNode(value));
+        } else {
+            *node = Some(last);
+
+            if self.nodes.len() == self.nodes.capacity() {
+                self.nodes.reserve(Self::NODES_CAPACITY);
             }
+
+            self.nodes.push(Node {
+                value,
+                left: None,
+                right: None,
+            });
         }
 
         Ok(())
     }
 
     /// Gets a reference to the value located at the end of the path.
     ///
-    /// This function traverses the tree from this node (`self`) until reaching the
+    /// This function traverses the tree from the root node until reaching the
     /// node at the end of the path. It returns [None], if the node is
     /// unreachable or nonexistent. Otherwise, it returns a reference to the
     /// value stored in the node.
     pub fn get(&self, path: &Path) -> Option<&V> {
-        let mut node = self;
+        let mut node = self.root;
```

Comment:
nit: this tree-traversal code is more-or-less duplicated several times (`insert`, `get`, and `get_node`); can we factor this out to a single implementation that we call into at each location that needs to traverse a tree?

Reply:
I expect we will need two traversal routines, one with & references and one with &mut references.

Reply:
I factored out the traversal, let me know what you think. Also due to mutability we need to (almost) duplicate code.